### PR TITLE
Ownership transferring refactoring

### PR DESCRIPTION
**Pull request by PhanLe1010**
1. When the manager pod on the owner node of the CR is missing, we switch
the ownerID of the CR to a different node
2. Don’t cleaning up CRs base on the current ownerID. The convention is 
that ownerID only indicates which node should watch the CR. We shouldn’t 
make any other logic depend on the ownerID. This way we can freely transfer
ownership
3. Make sure that when the engine/instance manager CR is transfered to 
new ownerID, the old owner must stop monitoring them. Also, remove 
engine/instance manager monitoring from the monitoring map of engine/instance 
manager controller when stop monitoring the engine. This allow the old
owner to restart the monitoring when the CRs’ owner is transfered back to it
4. Don’t clear the engine status when switching ownership. The new owner 
will take care of continuing to update the engine. This create a smooth
owner transferring.
5. Modify the isResponsibleFor() in engine, volume, engine image controller 
so that the current owner can continue to be the owner if the preferred 
owner doesn’t have engine image deployed.
6. Make sure the replica/engine instance is able to stop when the IM is 
in error state
7. The controller should at least look at each resource to determine
whether it's responsible for it or not, it's no the enqueue functions's
responsibility to make that determination
8. Allow attaching volume if there is engine image ready on at least one of 
the replicas' nodes of the volume.  This allow volumes to be attached/auto 
reattached when there are replicas on failed/non-exist nodes while still make
sure not to fail all of the replicas

longhorn/longhorn#2377
longhorn/longhorn#2329

### CODE REVIEW

File: controller/share_manager_controller.go
Comment (shuo-wu) [inline controller/share_manager_controller.go:294]: Why do we need to remove this?

File: controller/share_manager_controller.go
Comment (shuo-wu) [inline controller/share_manager_controller.go:496]: `&& volume.Spec.NodeID == ""`

File: controller/share_manager_controller.go
Comment (shuo-wu) [inline controller/share_manager_controller.go:483]: I am not sure if this part can be replaced. Since we want to make sure sm.Status.OwnerID == volume.Spec.NodeID == volume.Status.CurrentNodeID == pod.Spec.NodeName in this controller. Checking the node down only may be not enough.

File: manager/engineimage.go
Comment (shuo-wu) [inline manager/engineimage.go:120]: Do we need these? I think you can directly call `m.ds.CheckEngineImageReadyOnAllVolumeReplicas`

**Review by shuo-wu — CHANGES_REQUESTED** [review]
1. I have some doubts about function`isResponsibleFor()`:
    1. For the engine controller, it seems that we need to verify if the image the engine is using is on the node. Currently this function will blindly verify the default image.
    2. If the preferred owner has the ready engine image, can the current owner actively give up the ownership?
    3. It's better to list the strategy with comments. It's harder to understand this part...

File: controller/share_manager_controller.go
Comment (PhanLe1010) [reply controller/share_manager_controller.go:294]: We don't want to clean up the IM pod when transfer ownership to a new node. Doing so would disrupt the current running workloads.

File: controller/share_manager_controller.go
Comment (PhanLe1010) [reply controller/share_manager_controller.go:483]: IMHO,  we only need `volume.Spec.NodeID` == `pod.Spec.NodeName` so that the share manager pod can access the underlying Longhorn volume on the same node. The `sm.Status.OwnerID` and `volume.Status.CurrentNodeID` should be designed so that they can be moved freely to any controller node without disrupting the current running workloads that use the RWX volume. How do you think?
cc @joshimoo

File: manager/engineimage.go
Comment (PhanLe1010) [reply manager/engineimage.go:120]: You are right.  We don't need them. Fixed. Thank you

File: controller/share_manager_controller.go
Comment (PhanLe1010) [reply controller/share_manager_controller.go:496]: Thank you. Fixed

File: controller/engine_controller.go
Comment (PhanLe1010) [reply controller/engine_controller.go:601]: Can you explain why we need the to add `monitorNodeID`? Since the `ec.logger` already has the `controllerID`

File: controller/engine_controller.go
Comment (PhanLe1010) [reply controller/engine_controller.go:630]: My current solution for engine monitoring is: 
There are 3 goroutines: the main sync goroutine, the monitoring goroutine, and the stop monitoring goroutine. 
There are 2 ways to stop monitoring an engine:
1.  The main sync goroutine closes the stopCh -> the monitoring goroutine closes the monitoringRemoveCh -> the stop monitoring goroutine removes the engine RS from the `ec.engineMonitorMap` -> Done
1. The monitoring goroutine volunteer to stop monitoring engine, it closes the monitoringRemoveCh -> the stop monitoring goroutine removes the engine RS from the `ec.engineMonitorMap` -> Done

Your suggested solution for backing image monitoring is:
There are 3 goroutines: the main sync goroutine, the monitoring goroutine, and  the [notification update](https://github.com/shuo-wu/longhorn-manager/blob/35f06eba89ae87e0ad8e67b245b74c1ee2e1ac4c/controller/backing_image_manager_controller.go#L1093) goroutine
There are 2 ways to stop monitoring a backing image:
1.  The main sync goroutine closes the stopCh and removes the backing image RS from the `ec.engineMonitorMap` -> the monitoring goroutine returns ->  the notification update goroutine returns -> Done
1. The notification update goroutine volunteers to stop monitoring backing image CR, it closes the stopCh and  removes the engine RS from the `ec.engineMonitorMap`-> the monitoring goroutine returns -> Done

The suggested solution wouldn't work if the monitoring goroutine wants to volunteer to stop monitoring the backing image. For example, if something is wrong inside the monitoring goroutine while it does [pollAndUpdateBackingImageFileMap](https://github.com/shuo-wu/longhorn-manager/blob/35f06eba89ae87e0ad8e67b245b74c1ee2e1ac4c/controller/backing_image_manager_controller.go#L1147) (i.e., the backing image's owner has moved to a new node and doesn't need to be monitored on this node), the monitoring goroutine cannot volunteer to stop monitoring the backing image. If we try to call `m.monitorStopper(m.Name)` inside `pollAndUpdateBackingImageFileMap()`, it will be blocked forever because we send [stopCH <- struct{}{}](https://github.com/shuo-wu/longhorn-manager/blob/35f06eba89ae87e0ad8e67b245b74c1ee2e1ac4c/controller/backing_image_manager_controller.go#L1066) but no goroutine would take it out.

For engine monitoring, we need the stop monitoring goroutine to be able to volunteer to stop monitoring the engine if something is wrong while doing [sync](https://github.com/longhorn/longhorn-manager/blob/f12923c00c2aab6a21b52bf311393337814902c9/controller/engine_controller.go#L686). Therefore, I think the current implementation is ok.

How do you think?

File: controller/engine_controller.go
Comment (PhanLe1010) [reply controller/engine_controller.go:482]: Thank you! Fixed

File: controller/engine_controller.go
Comment (shuo-wu) [reply controller/engine_controller.go:581]: Oh, sorry. Then we don't need it

File: controller/engine_controller.go
Comment (shuo-wu) [reply controller/engine_controller.go:610]: Make sense. As a receiver, the monitor goroutine cannot send a stop signal.

File: controller/engine_controller.go
Comment (shuo-wu) [reply controller/engine_controller.go]: The map cleanup for the not running instance managers is missed.

File: controller/share_manager_controller.go
Comment (shuo-wu) [reply controller/share_manager_controller.go:483]: I am confused about the share manager state and the node ID issue now:
1. Can `sm.status.ownerID` be different from `volume.Spec.NodeID == pod.Spec.NodeName` while the related RWX volume is running?
2. Then when will the share manager become state unknown?

**Comment by shuo-wu** [comment]
> I have some doubts about function`isResponsibleFor()`:
> 
> 1. For the engine controller, it seems that we need to verify if the image the engine is using is on the node. Currently this function will blindly verify the default image.
> 2. If the preferred owner has the ready engine image, can the current owner actively give up the ownership?
> 3. It's better to list the strategy with comments. It's harder to understand this part...

You may miss this comment

File: controller/engine_controller.go
Comment (shuo-wu) [reply controller/engine_controller.go:610]: And I may need a similar logic in the backing image manager controller to actively stop monitoring in case of ownerID suddenly change.

Comment (shuo-wu) [reply controller/engine_controller.go:610]: BTW, can you rename it to avoid misunderstanding? e.g., `monitorVoluntaryStopCh`

**Comment by PhanLe1010** [comment]
Regard to https://github.com/longhorn/longhorn-manager/pull/854#pullrequestreview-624859923
> For the engine controller, it seems that we need to verify if the image the engine is using is on the node. Currently this function will blindly verify the default image.

Good catch! Thank you. I fixed it

> If the preferred owner has the ready engine image, can the current owner actively give up the ownership?

Yes. From the condition
```
(currentNodeEngineAvailable && isResponsible) ||
		(currentNodeEngineAvailable && !preferredOwnerEngineAvailable && ec.controllerID == e.Status.OwnerID) ||
		(currentNodeEngineAvailable && !preferredOwnerEngineAvailable && !currentOwnerEngineAvailable), nil
```
When the preferred owner has the ready engine image, on the current owner node, we will have:
* `isResponsible` is `false` because the preferred owner is up and running
*  `!preferredOwnerEngineAvailable ` is `false` since the preferred owner has the engine image
So the function return `false` and the current owner gives up the ownership

> It's better to list the strategy with comments. It's harder to understand this part...

I added some comments. Thank you. Yeah, the logic here is complicated because this function must make sure that it will output one and only one node when it is running on any node at any given time.

File: controller/share_manager_controller.go
Comment (PhanLe1010) [reply controller/share_manager_controller.go:483]: > Can sm.status.ownerID be different from volume.Spec.NodeID == pod.Spec.NodeName while the related RWX volume is running?

Yes, it can from the code perspective. I have also tested the scenario 

> Then when will the share manager become state unknown?

It never becomes unknown from the code perspective.  We also mark it as `error` when things went wrong. I think it is ok.

File: controller/engine_controller.go
Comment (PhanLe1010) [reply controller/engine_controller.go]: Thank you! I changed `monitoringRemoveCh ` to `monitorVoluntaryStopCh `

File: controller/instance_manager_controller.go
Comment (joshimoo) [inline controller/instance_manager_controller.go:279]: you can use `getLoggerForInstanceManager`

File: controller/engine_controller.go
Comment (joshimoo) [inline controller/engine_controller.go:410]: You can use `getLoggerForEngine`

File: controller/instance_manager_controller.go
Comment (joshimoo) [inline controller/instance_manager_controller.go:753]: use `getLoggerForInstanceManager`

File: controller/instance_manager_controller.go
Comment (joshimoo) [inline controller/instance_manager_controller.go:1140]: consider inverting this `if !needUpdate { continue }`
reduces nesting.

File: controller/kubernetes_configmap_controller.go
Comment (joshimoo) [inline controller/kubernetes_configmap_controller.go:36]: just remove the comment, doesn't really add anything :)

File: controller/replica_controller.go
Comment (joshimoo) [inline controller/replica_controller.go:277]: consider removing the comment, doesn't really add anything that is not expressed by.
We are not responsible for this :)

File: datastore/longhorn.go
Comment (joshimoo) [inline datastore/longhorn.go:979]: I feel like it's better to specify the return values explicitly instead of relying on the default values of the named params. In this case you also probably don't need the named params, since it's obvious from the usage :)

File: datastore/longhorn.go
Comment (joshimoo) [inline datastore/longhorn.go:991]: Consider whether you need the named params.

File: manager/volume.go
Comment (joshimoo) [inline manager/volume.go:322]:
```suggestion
		return nil, fmt.Errorf("cannot attach volume %v because the engine image %v is not deployed on at least one of the the replicas' nodes or the node that the volume is going to attach to", v.Name, v.Spec.EngineImage)
```

File: controller/volume_controller.go
Comment (joshimoo) [inline controller/volume_controller.go:2554]: You might want to use variables like this, similar to the ControllerIsResponsibleFor, might make it easier to digest what's going on :)

```
	isPreferredOwner := currentNodeEngineAvailable && isResponsible
	continueToBeOwner := currentNodeEngineAvailable && !preferredOwnerEngineAvailable && vc.controllerID == v.Status.OwnerID
	requiresNewOwner := currentNodeEngineAvailable && !preferredOwnerEngineAvailable && !currentOwnerEngineAvailable
	return isPreferredOwner || continueToBeOwner || requiresNewOwner
```

**Review by joshimoo — COMMENTED** [review]
Some feedback below, I will now have a deeper look at the engine_controller, instance_manager_controller, share_manager_controller will submit another review for that part.

**Comment by joshimoo** [comment]
Hey, can you take a look at the following code to see how I use a cancellable context for monitoring.
And see if this might be a better implementation instead of the raw stopChannel + voluntaryStopChannel?
The actual code details are not important for the example.

link to the relevant places below:
- https://github.com/joshimoo/longhorn-manager/blob/aa8f26dd0d18d55b68e10dde5fb2ad2045ac56c2/shareapi/monitor.go#L42 (cancellable context)
- https://github.com/joshimoo/longhorn-manager/blob/aa8f26dd0d18d55b68e10dde5fb2ad2045ac56c2/shareapi/monitor.go#L65-L66 (2 go routines)
- https://github.com/joshimoo/longhorn-manager/blob/aa8f26dd0d18d55b68e10dde5fb2ad2045ac56c2/shareapi/monitor.go#L70-L71 (close function that can be called externally, that calls the cancel function)

context evaluation:
- https://github.com/joshimoo/longhorn-manager/blob/aa8f26dd0d18d55b68e10dde5fb2ad2045ac56c2/shareapi/monitor.go#L101
- https://github.com/joshimoo/longhorn-manager/blob/aa8f26dd0d18d55b68e10dde5fb2ad2045ac56c2/shareapi/monitor.go#L123

Have a look at the enable/disable/hasMonitor functions (controller)
- https://github.com/joshimoo/longhorn-manager/blob/0fde5608ea2985829b1e1a5e2a82b22aad37c03f/controller/share_manager_controller.go#L35-L68

Usage of the monitor (this one is not important since it might require to much changes to the monitor for v1.1.1)
In General our monitoring is kinda flawed, since we have the reconcile loop directly call a grpc service instead of the monitor keeping state and ensuring a requeue of the elements
If you are curious how monitoring should work, have a look at the syncCallback usage :slightly_smiling_face:
- https://github.com/joshimoo/longhorn-manager/blob/0fde5608ea2985829b1e1a5e2a82b22aad37c03f/controller/share_manager_controller.go#L415

File: controller/share_manager_controller.go
Comment (joshimoo) [inline controller/share_manager_controller.go:793]: isCurrentNodeSchedulable is only relevant, if the pod currently does not exists (share-manager-state == error), since you use the pod.spec.NodeName as preferred owner. If the pod already exists the schedulable status is irrelevant.

File: controller/share_manager_controller.go
Comment (joshimoo) [inline controller/share_manager_controller.go:781]: You can rename this to isResponsibleFor same as the other controllers, since we are now doing the same more or less :)

File: controller/share_manager_controller.go
Comment (joshimoo) [reply controller/share_manager_controller.go:483]: Using status.Owner as condition was a simplification for volume failure handling.
It's not strictly required that the owner for the crd is the same node as the actual pod.
So this change is fine, but comes with the caveat that it adds a couple more isNodeDown checks  :)

But overall I agree, it's okay to decouple CRD processing from pod/volume ownership.

**Review by joshimoo — COMMENTED** [review]
Left some notes about the share-manager-controller, I think your changes in that regard are okay.
Can you check on my comment in regard to the monitoring with cancellable context to see if that can simplify/make the monitoring more robust?

**Comment by PhanLe1010** [comment]
Regarding to https://github.com/longhorn/longhorn-manager/pull/854#issuecomment-812101940

> Usage of the monitor (this one is not important since it might require to much changes to the monitor for v1.1.1)
In General our monitoring is kinda suboptimal, since we have the reconcile loop directly call a grpc service instead of the monitor keeping state and ensuring a requeue of the elements
If you are curious how monitoring could be improved, have a look at the syncCallback usage

Thank you for the suggestion. I agree that it is a better design and easier to read. Yeah, it would require  a big change, we can consider refactoring the monitoring logic of the engine, instance manager in a different issue. I will create an issue to track the effort. 

> And see if this might be a better implementation instead of the raw stopChannel + voluntaryStopChannel?

Until we do the above refactoring, I don't see how the cancellable context will help to make the code cleaner comparing to the raw stopChannel + voluntaryStopChannel. I think we can wait until we do refactoring and use the cancellable context.

How do you think?

File: controller/engine_image_controller.go
Comment (PhanLe1010) [reply controller/engine_image_controller.go:817]: You are right. This change is not logically necessary. I just want to make the code format to be similar to other controllers.

File: controller/instance_manager_controller.go
Comment (PhanLe1010) [reply controller/instance_manager_controller.go:279]: Thank you. Fixed

File: controller/instance_manager_controller.go
Comment (PhanLe1010) [reply controller/instance_manager_controller.go:753]: Thank you. Fixed

File: controller/instance_manager_controller.go
Comment (PhanLe1010) [reply controller/instance_manager_controller.go:1140]: Thank you. Fixed

File: controller/kubernetes_configmap_controller.go
Comment (PhanLe1010) [reply controller/kubernetes_configmap_controller.go:36]: Thank you. Fixed

File: controller/replica_controller.go
Comment (PhanLe1010) [reply controller/replica_controller.go:277]: Thank you. Fixed

File: controller/share_manager_controller.go
Comment (PhanLe1010) [reply controller/share_manager_controller.go:781]: Thank you. Fixed

File: controller/share_manager_controller.go
Comment (PhanLe1010) [reply controller/share_manager_controller.go:793]: I see. But I think it is still ok that we switch to a schedulable node even if the pod already exists because:
1. Ownership transferring has a very small cost now
2. The next time the sm stops, it will have to switch to the new node anyways
How do you think?

File: controller/volume_controller.go
Comment (PhanLe1010) [reply controller/volume_controller.go:2554]: Thank you. Fixed

File: datastore/longhorn.go
Comment (PhanLe1010) [reply datastore/longhorn.go:979]: Thank you. Fixed

File: datastore/longhorn.go
Comment (PhanLe1010) [reply datastore/longhorn.go:991]: Thank you. Removed

File: controller/engine_controller.go
Comment (PhanLe1010) [reply controller/engine_controller.go:410]: Thank you. Fixed

File: manager/volume.go
Comment (PhanLe1010) [reply manager/volume.go:322]: Thank you. Fixed

File: controller/engine_controller.go
Comment (shuo-wu) [reply controller/engine_controller.go]: OK, I see

File: controller/instance_manager_controller.go
Comment (joshimoo) [inline controller/instance_manager_controller.go:282]: don't need the comment.

File: controller/engine_image_controller.go
Comment (joshimoo) [reply controller/engine_image_controller.go:816]: Sounds reasonable, do you want to use variables like we do for the other controllers.
isPreferredOwner, continueToBeOwner, requiresNewOwner 
I find with the named variables it makes the intentions/conditions very easy to understand.

**Review by joshimoo — APPROVED** [review]
LGTM once the nits are addressed this can be merged.
I will let @shuo-wu merge it unless I receive a notification that it's good to merge from him/phan.

File: controller/share_manager_controller.go
Comment (PhanLe1010) [reply controller/share_manager_controller.go:793]: Thank you. Fixed

File: controller/instance_manager_controller.go
Comment (PhanLe1010) [reply controller/instance_manager_controller.go]: Thank you. Fixed

**Review by innobead — APPROVED** [review]
LGTM

Any other concerns @joshimoo @shuo-wu? Otherwise we can merge

**Review by shuo-wu — APPROVED** [review]

**Pull request merged**
